src/coreclr/jit/helperexpansion.cpp

```diff
@@ -0,0 +1,1043 @@
+    while ((newFirstStmt != nullptr) && (newFirstStmt != stmt))
+    {
+        fgMorphStmtBlockOps(block, newFirstStmt);
+        newFirstStmt = newFirstStmt->GetNextStmt();
+    }
+
+    GenTreeLclVar* threadStaticBlockLcl = nullptr;
+
+    // Grab a temp to store result (it's assigned from either fastPathBb or fallbackBb)
+    if (threadStaticBlockLcl == nullptr)
+    {
+        // Define a local for the result
+        unsigned threadStaticBlockLclNum         = lvaGrabTemp(true DEBUGARG("TLS field access"));
+        lvaTable[threadStaticBlockLclNum].lvType = TYP_I_IMPL;
+        threadStaticBlockLcl                     = gtNewLclvNode(threadStaticBlockLclNum, call->TypeGet());
+
+        *callUse = gtClone(threadStaticBlockLcl);
+
+        fgMorphStmtBlockOps(block, stmt);
+        gtUpdateStmtSideEffects(stmt);
+    }
+
+    GenTree* typeThreadStaticBlockIndexValue = call->gtArgs.GetArgByIndex(0)->GetNode();
+
+    void** pIdAddr = nullptr;
+
+    size_t   tlsIndexValue = (size_t)threadStaticBlocksInfo.tlsIndex.addr;
+    GenTree* dllRef        = nullptr;
+
+    if (tlsIndexValue != 0)
+    {
+        dllRef = gtNewIconHandleNode(tlsIndexValue * TARGET_POINTER_SIZE, GTF_ICON_TLS_HDL);
+    }
+
+    // Mark this ICON as a TLS_HDL, codegen will use FS:[cns] or GS:[cns]
+    GenTree* tlsRef = gtNewIconHandleNode(threadStaticBlocksInfo.offsetOfThreadLocalStoragePointer, GTF_ICON_TLS_HDL);
+
+    tlsRef = gtNewIndir(TYP_I_IMPL, tlsRef, GTF_IND_NONFAULTING | GTF_IND_INVARIANT);
+
+    if (dllRef != nullptr)
+    {
+        // Add the dllRef to produce thread local storage reference for coreclr
+        tlsRef = gtNewOperNode(GT_ADD, TYP_I_IMPL, tlsRef, dllRef);
+    }
+
+    // Base of coreclr's thread local storage
+    GenTree* tlsValue = gtNewIndir(TYP_I_IMPL, tlsRef, GTF_IND_NONFAULTING | GTF_IND_INVARIANT);
+
+    // Cache the tls value
+    unsigned tlsLclNum         = lvaGrabTemp(true DEBUGARG("TLS access"));
+    lvaTable[tlsLclNum].lvType = TYP_I_IMPL;
+    GenTree* defTlsLclValue    = gtNewLclvNode(tlsLclNum, TYP_I_IMPL);
+    GenTree* useTlsLclValue    = gtCloneExpr(defTlsLclValue); // Create a use for tlsLclValue
+    GenTree* asgTlsValue       = gtNewAssignNode(defTlsLclValue, tlsValue);
+
+    // Create tree for "maxThreadStaticBlocks = tls[offsetOfMaxThreadStaticBlocks]"
+    GenTree* offsetOfMaxThreadStaticBlocks =
+        gtNewIconNode(threadStaticBlocksInfo.offsetOfMaxThreadStaticBlocks, TYP_I_IMPL);
+    GenTree* maxThreadStaticBlocksRef =
+        gtNewOperNode(GT_ADD, TYP_I_IMPL, gtCloneExpr(useTlsLclValue), offsetOfMaxThreadStaticBlocks);
+    GenTree* maxThreadStaticBlocksValue =
+        gtNewIndir(TYP_INT, maxThreadStaticBlocksRef, GTF_IND_NONFAULTING | GTF_IND_INVARIANT);
+
+    // Create tree for "if (maxThreadStaticBlocks < typeIndex)"
+    GenTree* maxThreadStaticBlocksCond =
+        gtNewOperNode(GT_LT, TYP_UINT, maxThreadStaticBlocksValue, gtCloneExpr(typeThreadStaticBlockIndexValue));
```

Comment:
Shouldn't it be just `TYP_INT` here? `TYP_UINT` on this node is a sort of invalid IR. If you want unsigned comparison you need to set `GTF_UNSIGNED` flag

Reply:
Changed to `TYP_INT`.